src/components/multi-select/multi-select.jsx

```diff
@@ -0,0 +1,316 @@
+import styles from './multi-select.css';
+
+const MultiSelect = forwardRef(function MultiSelect(props, ref) {
+  const [autocompleteValue, setAutocompleteValue] = useState('');
+  const autocompleteRef = useRef();
+  const backupRef = useRef();
+  const [expanded, setExpanded] = useState(false);
+  const selfRef = ref || backupRef;
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+  const [value, setValue] = useState(props.value || []);
+  const valueRefs = value.map(() => createRef());
+  const visibleOptions = getVisibleOptions();
+  const visibleOptionsRefs = visibleOptions.map(() => createRef());
+  const wrapperRef = useRef(null);
+
+  const ids = {
+    emptyMessage: `${props.id}-empty`,
+    label: `${props.id}-label`,
+    listbox: `${props.id}-listbox`,
+    textbox: `${props.id}-autocomplete`,
+  };
+
+  const classNames = {
+    container: styles.container,
+    formControlChildrenContainer: props.readOnly ? styles['form-control-children-container-readonly'] : styles['form-control-children-container'],
+    iconClear: props.readOnly ? styles['icon-clear-readonly'] : styles['icon-clear'],
+    iconsContainer: styles['icon-container'],
+    input: props.readOnly ? styles['input-readonly'] : styles.input,
+    noOptionsContainer: styles['no-options'],
+    popupContainer: styles['popup-container'],
+    tagList: styles['tag-list'],
+    ...props.classNames,
+  };
+
+  const dropdownContents = () => {
+    if (!expanded) {
+      return undefined;
+    }
+
+    if (props.showLoader) {
+      return (
+        <Listbox
+          className={classNames.popupContainer}
+          id={ids.listbox}
+          labelledBy={ids.label}
+          refs={[]}
+        >
+          <Loader inline />
+        </Listbox>
+      );
+    }
+
+    if (visibleOptions.length === 0) {
+      return (<NoOptions className={classNames.noOptionsContainer} id={ids.emptyMessage} />);
+    }
+
+    return (
+      <Listbox
+        className={classNames.popupContainer}
+        id={ids.listbox}
+        labelledBy={ids.label}
+        refs={visibleOptionsRefs}
+      >
+        {props.listboxChildren ?
+          props.listboxChildren(visibleOptions, visibleOptionsRefs, onOptionSelect) :
+          visibleOptions.map((option, index) => (
+            <ListOption
+              key={`${props.id}-${props.optionIDGetter(option)}`}
+              onSelect={onOptionSelect}
+              ref={visibleOptionsRefs[index]}
+              value={props.optionIDGetter(option)}
+            >
+              {props.optionLabelGetter(option)}
+            </ListOption>
+          ))}
+      </Listbox>
+    );
+  };
+
+  function getOption(id) {
+    return props.options.find(option => props.optionIDGetter(option) === id);
+  }
+
+  function getVisibleOptions() {
+    return props.options
+      .filter(option => props.optionLabelGetter(option).toLowerCase().includes(autocompleteValue.toLowerCase()))
```

Comment:
Note: the use of `toLowerCase` here fixes what I _think_ is a bug/unintended consequence of the `CustomFieldInputMultiChoice` implementation. If we _don't_ want to do a case-insensitive match, I can remove this.

Reply:
Ah, yeah, I think so. Let's remove the case-insensitivity if my comment below checks out.

Reply:
Case insensitivity for filtering options was kept after discussion, when the new `filterOptions` prop is `true`.